flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackendBuilder.java

```diff
@@ -498,20 +505,36 @@ private RocksDBRestoreOperation getRocksDBRestoreOperation(
         RocksDBSnapshotStrategyBase<K, ?> checkpointSnapshotStrategy;
         if (enableIncrementalCheckpointing) {
             checkpointSnapshotStrategy =
-                    new RocksIncrementalSnapshotStrategy<>(
-                            db,
-                            rocksDBResourceGuard,
-                            keySerializerProvider.currentSchemaSerializer(),
-                            kvStateInformation,
-                            keyGroupRange,
-                            keyGroupPrefixBytes,
-                            localRecoveryConfig,
-                            cancelStreamRegistry,
-                            instanceBasePath,
-                            backendUID,
-                            materializedSstFiles,
-                            lastCompletedCheckpointId,
-                            numberOfTransferingThreads);
+                    injectRocksDBStateUploader != null
+                            ? new RocksIncrementalSnapshotStrategy<>(
```

Comment:
There are two fields now in builder: `injectRocksDBStateUploader` and `numberOfTransferingThreads`. But if the former is set then the latter is ignored.

I think the more robust solution would be to  throw an exception in `setNumberOfTransferingThreads` if the uploader is not null. Otherwise, we'd have to close the old uploader, which is not something I'd expect from a builder setter method.

WDYT?

Reply:
I think your ideas are more rigorous. please wait for my fix